File: Data Structures & Algorithms/balanced-binary-tree/submission-11.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBalanced(TreeNode root) {
        int [] result = isValid(root);
        return result[0] == 1 ? true : false;
    }

    public int [] isValid(TreeNode root){
        if(root == null){
            return new int [] {1, 0};
        }

        int [] leftResult = isValid(root.left);
        int [] rightResult = isValid(root.right);

        boolean isCurrentValid = (leftResult[0] == 1 && rightResult[0] == 1 && Math.abs(leftResult[1] - rightResult[1]) <= 1);

        int maxDepth = Math.max(leftResult[1], rightResult[1]) + 1;


        return new int []{isCurrentValid ? 1 : 0, maxDepth};
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-11.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        current_diameter = [0]
        self.dfs(root, current_diameter)

        return current_diameter[0]

    def dfs(self, root, current_diameter):
        if root is None:
            return 0
        
        left_value = self.dfs(root.left, current_diameter)
        right_value = self.dfs(root.right, current_diameter)

        current_diameter[0] = max(current_diameter[0], left_value + right_value)

        return max(left_value, right_value) + 1







        
        
        
        

        

        